src/mrinufft/io/pulseq.py

```diff
@@ -0,0 +1,255 @@
+"""Pulseq Trajectory Reader and writers.
+
+Reads Pulseq `.seq` files to extract k-space trajectory and other parameters. It
+also provides functionality to create pulseq block and shape objects to
+facilitate the integration of arbitrary k-space trajectories into Pulseq
+sequences. Requires the `pypulseq` package to be installed.
+"""
+
+import warnings
+from types import SimpleNamespace
+import numpy as np
+import pypulseq as pp
+from numpy.typing import NDArray
+from .utils import prepare_trajectory_for_seq
+from mrinufft.trajectories.utils import KMAX
+
+
+def read_pulseq_traj(seq, traj_delay=0.0, grad_offset=0.0):
+    """Extract k-space trajectory from a Pulseq sequence file.
+
+    The sequence should be a valid Pulseq `.seq` file, with arbitrary gradient
+    waveforms, which all have the same length.
+
+    Note
+
+    Parameters
+    ----------
+    sequence : pulseq Sequence, or Path to the Pulseq `.seq` file.
+
+    Returns
+    -------
+    description : dict
+        the [DESCRIPTION] block from the sequence file.
+    np.ndarray
+        The k-space trajectory as a numpy array of shape (n_shots, n_samples, 3),
+        where the last dimension corresponds to the x, y, and z coordinates in k-space.
+
+    """
+    if not isinstance(seq, pp.Sequence):
+        filename = seq
+        seq = pp.Sequence()
+        seq.read(filename)
+
+    kspace_adc, _, t_exc, t_refocus, t_adc = seq.calculate_kspace()
+
+    if not t_adc:
+        raise ValueError(
+            "The sequence does not contain any ADC events. "
+            "Please ensure that the sequence has ADC events defined."
+        )
+    # split t_adc with t_exc and t_refocus, the index are then used to split kspace_adc
+    FOV = seq.get_definition("FOV")
+    t_splits = np.sort(np.concatenate([t_exc, t_refocus]))
+    idx_prev = 0
+    kspace_shots = []
+    for t in t_splits:
+        idx_next = np.searchsorted(t_adc, t, side="left")
+        if idx_next == idx_prev:
+            continue
+        kspace_shots.append(kspace_adc[:, idx_prev:idx_next].T)
+        if idx_next == kspace_adc.shape[1] and t > t_adc[-1]:  # last useful point
+            break
+        idx_prev = idx_next
+    if idx_next < kspace_adc.shape[1]:
+        kspace_shots.append(kspace_adc[:, idx_next:].T)  # add remaining gradients.
+    # convert to KMAX standard.
+    kspace_shots = np.ascontiguousarray(kspace_shots) * KMAX * 2 * np.asarray(FOV)
+    return seq.definitions, kspace_shots
+
+
+def _check_timings(seq):
+    # Check whether the timing of the sequence is correct
+    ok, error_report = seq.check_timing()
+    if ok:
+        return None
+    else:
+        warnings.warn("Timing check failed. Error listing follows:" + str(error_report))
```

Comment:
Why a separate function really?

Reply:
Cause I am planning to add other sequences (like a 2D GRE for instance) this would be common to all of them.